### DISCUSSION TITLE
solid particle temperature distribution interpolation in halfcell DFN

### DISCUSSION THREAD

**Discussion by xyyrly** [pybamm-team/PyBaMM#3289] [Q&A]
Hi Team,
I am trying to set an initial temperature distribution for positive electrode particles in DFN model. The following was what I did:
```python3
x_var = pybamm.SpatialVariable("x", domain=["positive electrode"])
x_disc = disc.process_symbol(x_var)
x_eva = x_disc.evaluate()
r_p_var = pybamm.SpatialVariable("r", domain=["positive particle"])
r_p_disc = disc.process_symbol(r_p_var)
r_p_eva = r_p_disc.evaluate()
xxx = (r_p_eva[:,0],x_eva[:,0])
r_pp=pybamm.standard_spatial_vars.r_p
x_pp= pybamm.PrimaryBroadcast(pybamm.standard_spatial_vars.x_p, "positive particle")

Tsp=np.ones((20,20))*298.15 (for example)
T_s_p=pybamm.Interpolant(xxx, Tsp, (r_pp,x_pp), interpolator="linear") 
N_s_p = -param.p.prim.D(c_s_w, T_s_p) * pybamm.grad(c_s_w)
    model.rhs[c_s_w] = -pybamm.div(N_s_p)
    model.boundary_conditions[c_s_w] = {
        "left": (pybamm.Scalar(0), "Neumann"),
        "right": (
            -j_p / (param.F * pybamm.surf(param.p.prim.D(c_s_w, T_s_p))),
            "Neumann",
        ),
    }
```

There is an error when doing "param.p.prim.D(c_s_w, T_s_p) * pybamm.grad(c_s_w)": 
![1692667474(1)](https://github.com/pybamm-team/PyBaMM/assets/117139564/c7a28ff8-0eaa-4288-b37f-4b609e6f3ae2)
I checked the shape of D(c_s_w, T_s_p) and grad(c_s_w) after calculating 1 time step, the first is (20,20,2), the second is (21,20,2), it looks correct..

One weird thing is weeks ago I used the same interpolation in a full DFN model for LIB, that time it worked smoothly but now I changed to halfcell DFN for LMB and it failed....

Thank you very much for your time!

**Comment by xyyrly** [comment]
Oh now the interpolation in the original full DFN model for LIB failed too. Are there some related changes in pybamm?

**Comment by brosaplanella** [comment]
The `param.p.prim.D(c_s_w, T_s_p) * pybamm.grad(c_s_w)` should work, but it is not clear which line is throwing the error. Can you share a minimum working example and the full error log you are getting?
